### PR TITLE
Esilky  unit tests#378

### PR DESCRIPTION
**Pull request by esilky**
This:
* adds a 'main' package for the 'Driver' class to be in (to conform to the Java spec and to allow for unit testing).
* makes the <<default-package>>.Driver.main method create a Driver object and use it to run the client.  This allows the start scripts to continue to run correctly without specifying a package (helpful in general).
* adds an initial collection of JUnit tests
* slightly refactors some classes to allow them to be tested
* adds some useful methods to the StatusIndicator interface
* fixes a problem with the StatusConsole class that wasn't found until implementing the unit tests
* adds JUnit foundation to the project in the form of:
** JUnit test categories (these allow tests to be categorized as 'general', 'slow-running', other...
** JUnit Test Suite base class - where test classes can be registered to be included in Test Suites
** JUnit Test Suite All - runs all of the unit tests registered in the base class
** JUnit Test Suite General - runs all of the 'general' (non long-running) tests registered in the base class

These tests can be executed from within Eclipse by using the 'Run As - JUnit Test' right-click menu command on any of the test classes or the 'Suite' classes (in the _junitTests folder).

<< ToDo - implement running the tests from a Maven command as a separate change >>

### CODE REVIEW

**Comment by esilky** [comment]
Sorry if I'm doing this wrong... I've worked with SCC for decades, but I'm new to Git & GitHub.

This includes (or seems to include) changes from the previous two pull requests I submitted.  If that is correct - great!  If not, can you give me pointers on what I need to do to limit the 'pull' to just the changes in this branch. (other SCC systems I've worked with would only list the changes that were made in this branch)

**Comment by esilky** [comment]
I have some changes for enabling the Maven project to run the tests, but I'm going to wait to create an issue and branch for it until these get approved/merged.

**Comment by erezbi** [comment]
@esilky I think once we merge the previous PRs, and you pull from master into this branch, then only the new changes should be visible here.
Also, we create a branch for each task. So each PR should have little changes in it, and probably they will not conflict. You should work based on master in the different branches that you do, and not based on each of your branches, if possible. Does this make sense?

**Comment by esilky** [comment]
@erezbi - that makes sense, I think the problem might be with the time-zone delay. When I submit a pull, I then move on to the next task.  To do that, I create a new branch - but the previous branch changes are included until they are accepted and merged (at least that is what it seems). So, if I want to submit the next task, it includes the previous changes (or seems to).

Again, I have worked with almost a dozen SCC Systems, but not very much with Git (and it is not intuitive - in my opinion).

**Comment by esilky** [comment]
@erezbi - If I work from master, then it does not include the changes I made to-date. In some cases that should be fine - but if I've made changes to (say) logging (Log4J) that have not been accepted yet, but I want to use that Log4J functionality for logging of 'performance' I can't simply work from master.

**Comment by erezbi** [comment]
@esilky you are right. We usually are able to work based on master. You can work based on other branch, I think that if it accepted before the second one, then there will not be a conflict. But until the first one is accepted, the second will show more changes than really exist there.

**Comment by esilky** [comment]
@erezbi Okay, that makes sense.  I was just a bit confused why all of these changes were showing up.

**Comment by esilky** [comment]
@erezbi @arivatibm - Is there anything needed from me to get this merged?

File: src/main/java/wa/client/Client.java [hunk 1]
```diff
@@ -304,7 +305,7 @@ public void run() {
                     }
                     if (ServerConnectionStatus.NOTCONNECTED == getServerConnectionStatus()) {
                     		// IAMAccessToken should be retrieved from the IAM service by providing it with your cloud API Key (based on your IBM ID)
-                        IAMAccessToken = getIAMAccessToken(IAMAPIKey);
+                        IAMAccessToken = getIAMAccessToken(keyIAMAPI);
```
Comment (arivatibm) [inline src/main/java/wa/client/Client.java]: same comment as in the previous pull request - change name to iamApiKey

**Review by arivatibm — CHANGES_REQUESTED** [review]

**Comment by arivatibm** [comment]
@esilky Please ask Ramesh to test this after you are done with the code fix

**Comment by arivatibm** [comment]
QA testing by Ramesh
I've asked him to do it just now
 
                                                                        Regards,
                         
                        Ari Volcoff                        Research Staff Member                        Watson Assistant Solutions , IoT And Wearables                        IBM - Haifa Research Lab                                                                         
                                                                                                                                                        Mobile: +972-58-4168357 | Phone: +972-4-8281319                        E-mail: ARIV@il.ibm.com                        LinkedIn: https://il.linkedin.com/in/avolcoff                                                
 
 
----- Original message -----From: Ed Silky <notifications@github.com>To: Watson-Personal-Assistant/AudioClientSampleCodeJava <AudioClientSampleCodeJava@noreply.github.com>Cc: arivatibm <ariv@il.ibm.com>, Mention <mention@noreply.github.com>Subject: Re: [Watson-Personal-Assistant/AudioClientSampleCodeJava] Esilky unit tests#378 (#4)Date: Wed, Jun 6, 2018 9:26 PM 
@erezbi @arivatibm - Is there anything needed from me to get this merged?
—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by arivatibm** [comment]
@esilky  told Ramesh he needs to make some additional fixes 
@erezbi FYI

File: src/main/java/wa/client/Client.java (hunk 1 above)
Comment (esilky) [reply src/main/java/wa/client/Client.java]: @arivatibm Changed to iamApiKey per request.

**Comment by esilky** [comment]
@rameshupadhyaper This is tested by:
- opening the project in Eclipse
- use Enterprise Explorer and open the tree at `testsrc/main/java : _junitTests`
- right click on `JUnitTestSuite_All.java` and select 'Run as > JUnit Test'
- open the JUnit view (might need to open from Window>Show View
- check that 5/5 tests ran with 0 errors and 0 failures (green bar)
- do the same for `JUnitTestSuite_General.java` and verify that 4/4 tests ran without errors/failures

**Comment by esilky** [comment]
@rameshupadhyaper please let @arivatibm know when you have finished your testing.

@arivatibm do you need to do anything to clear your review? I made the requested changes.

**Review by arivatibm — APPROVED** [review]

**Comment by arivatibm** [comment]
@esilky there are some merge conflicts - can you resolve them by merging the master into your feature branch?

**Comment by esilky** [comment]
@arivatibm @erezbi the latest master is merged in.  There shouldn't be any conflicts.

**Comment by erezbi** [comment]
@esilky I see you made more commits. Did @rameshupadhyaper test the PR after this last commit
(eb77668 above)?

**Comment by esilky** [comment]
These were part of the merge from master into the branch for this PR.
@rameshupadhyaper have you tested?

**Comment by rameshupadhyaper** [comment]
@esilky 
@erezbi 
Tested in window machine.
Working fine as per requirement.

**Pull request merged**